src/main/java/com/google/devtools/build/lib/actions/ResourceManager.java

```diff
@@ -527,59 +496,43 @@ boolean areResourcesAvailable(ResourceSet resources) throws NoSuchElementExcepti
 
-    if (usedCpu == 0.0
-        && usedRam == 0.0
-        && usedExtraResources.isEmpty()
-        && usedLocalTestCount == 0
-        && workerIsAvailable) {
+    if (usedResources.isEmpty() && usedLocalTestCount == 0) {
       return true;
     }
-    // Use only MIN_NECESSARY_???_RATIO of the resource value to check for
-    // allocation. This is necessary to account for the fact that most of the
-    // requested resource sets use pessimistic estimations. Note that this
-    // ratio is used only during comparison - for tracking we will actually
-    // mark whole requested amount as used.
-    double cpu = resources.getCpuUsage() * MIN_NECESSARY_CPU_RATIO;
-    double ram = resources.getMemoryMb() * MIN_NECESSARY_RAM_RATIO;
-    int localTestCount = resources.getLocalTestCount();
-
-    double availableCpu = availableResources.getCpuUsage();
-    double availableRam = availableResources.getMemoryMb();
-    int availableLocalTestCount = availableResources.getLocalTestCount();
 
-    double remainingRam = availableRam - usedRam;
+    int availableLocalTestCount = availableResources.getLocalTestCount();
+    if (!isAvailable(availableLocalTestCount, usedLocalTestCount, resources.getLocalTestCount())) {
+      return false;
+    }
 
-    // Resources are considered available if any one of the conditions below is true:
-    // 1) If resource is not requested at all, it is available.
-    // 2) If resource is not used at the moment, it is considered to be
-    // available regardless of how much is requested. This is necessary to
-    // ensure that at any given time, at least one thread is able to acquire
-    // resources even if it requests more than available.
-    // 3) If used resource amount is less than total available resource amount.
-    boolean cpuIsAvailable = cpu == 0.0 || usedCpu == 0.0 || usedCpu + cpu <= availableCpu;
-    boolean ramIsAvailable = ram == 0.0 || usedRam == 0.0 || ram <= remainingRam;
-    boolean localTestCountIsAvailable =
-        localTestCount == 0
-            || usedLocalTestCount == 0
-            || usedLocalTestCount + localTestCount <= availableLocalTestCount;
-    boolean extraResourcesIsAvailable = areExtraResourcesAvailable(resources);
-    return cpuIsAvailable
-        && ramIsAvailable
-        && extraResourcesIsAvailable
-        && localTestCountIsAvailable
-        && workerIsAvailable;
+    for (Map.Entry<String, Double> resource : resources.getResources().entrySet()) {
+      String key = resource.getKey();
+
+      // Use only MIN_NECESSARY_RATIO of the resource value to check for
+      // allocation. This is necessary to account for the fact that most of the
+      // requested resource sets use pessimistic estimations. Note that this
+      // ratio is used only during comparison - for tracking we will actually
+      // mark whole requested amount as used.
+      double requested = resource.getValue() * MIN_NECESSARY_RATIO.getOrDefault(key, 1.0);
```

Comment:
Could you define `1.0` in a static constant above together with `MIN_NECESSARY_RATIO` (e.g. `DEFAULT_RATIO`)? That'll make it clearer what happens if it's not specified in the `MIN_NECESSARY_RATIO` map.

Reply:
Done